WalletWasabi.Fluent/ViewModels/Wallets/Advanced/WalletCoins/WalletCoinsViewModel.cs

```diff
@@ -98,34 +102,37 @@ public WalletCoinsViewModel(WalletViewModel walletViewModel, IObservable<Unit> b
 		};
 
 		Source.RowSelection!.SingleSelect = true;
-
 	}
 
 	public FlatTreeDataGridSource<WalletCoinViewModel> Source { get; }
 
+	private IObservable<Unit> CoinsUpdated => _balanceChanged
+		.ToSignal()
+		.Merge(_walletViewModel
+			.WhenAnyValue(w => w.IsCoinJoining)
+			.ToSignal());
+
 	protected override void OnNavigatedTo(bool isInHistory, CompositeDisposable disposables)
 	{
 		base.OnNavigatedTo(isInHistory, disposables);
 
-		Observable.Merge(
-			_balanceChanged.Select(_ => Unit.Default),
-			_walletViewModel.WhenAnyValue(w => w.IsCoinJoining).Select(_ => Unit.Default))
-			.Subscribe(_ =>
-			{
-				Update();
-			});
-
-		disposables.Add(_coinsSourceList);
+		CoinsUpdated
+			.Select(_ => GetCoins())
+			.Subscribe(RefreshCoinsList)
+			.DisposeWith(disposables);
 	}
 
-	private void Update()
+	private ICoinsView GetCoins()
 	{
-		var coins = _walletViewModel.Wallet.Coins.Select(c => new WalletCoinViewModel(c));
+		return _walletViewModel.Wallet.Coins;
+	}
 
+	private void RefreshCoinsList(ICoinsView items)
+	{
 		_coinsSourceList.Edit(x =>
 		{
 			x.Clear();
-			x.AddRange(coins);
+			x.AddRange(items.Select(coin => new WalletCoinViewModel(coin)));
```

Comment:
I think here is an opportunity to improve the performance just by applying the difference instead of clear and add. But do not touch this now it is not a priority.